### DISCUSSION TITLE
Uncaught ReferenceError: process is not defined while upgrading to webpack 5

### DISCUSSION THREAD

**Discussion by prashantkn94** [webpack/webpack#17744] [Q&A]
This runtime error is thrown because partner team's code inside node_modules uses process. like below:

`const isTestEnv = process?.env?.SOME_FIELD !== undefined;`

1. My first question is why does webpack throw this error at all since the optional chaining is being used here?
When I try to fix the error by defining plugin, the error still remains the same

Snippet from my webpack plugin:

```
plugins: [
    new webpack.DefinePlugin({
        'process?.env?.JEST_WORKER_ID': undefined,
    }),
]
```
2. I cannot use a polyfill for this since the code changes will be required at partner team end because they are not assigning process from any import. they just use it like a global variable. what option do I have ?

**Comment by lucasrmendonca** [comment]
The problem is that Webpack errors out for unreferenced variables instead of assuming they'll be undefined. To fix this, define only the "process" global without the optional chaining:

`
plugins: [
    new webpack.DefinePlugin({
        'process': null,
    }),
]
`

Then, access it inside the code with the optional chaining operator:

`
const isTestEnv = process?.env?.SOME_FIELD !== undefined;
`
